Use galaxy_directory() util function to check if code is running from root or from packages

jdavcs opened the pull request:
Use existing utility function instead of reinventing the wheel. Initially brought up by @neoformit - i.e., that it would've been nice to have a function for this ...and apparently we've had one all along.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jdavcs committed:
Use galaxy_directory() instead of hacking the path

jdavcs committed:
Refactor test, use galaxy_directory()

jdavcs committed:
Refactor custom_templates, use galaxy_root()

jmchilton commented on test/unit/data/model/migrations/test_dbscript.py:
```diff
@@ -130,19 +131,8 @@ def dburl_from_config(config: Config) -> str:
-    return completed_process
-
-
-def in_packages() -> bool:
```
I mean... if in packages skip the test is what I'd do here. The package tests should pretend they don't have galaxy_directory() available because they're meant to be installed into a Python env 🤔. This test is fundamentally not testing anything about the Galaxy data package - it is testing scripts placed in Galaxy root directory and that would not be available in an installed Galaxy environment. I would also move this test to... the app package instead of the data package for that reason I think.

jdavcs replied:
That never occurred to me.. It makes perfect sense, of course. Will make the changes, thanks!

jmchilton commented on test/unit/data/model/migrations/test_migrations.py:
```diff
@@ -1122,13 +1123,7 @@ def db_state6_gxy_state3_tsi_no_sam(url_factory, metadata_state6_gxy_state3_tsi_
 @pytest.fixture(autouse=True)
 def legacy_manage_db(monkeypatch):
     def get_alembic_cfg():
-        path = os.path.join(os.path.dirname(__file__), os.pardir, os.pardir, os.pardir, os.pardir, os.pardir)
```
For something like this you can use something like... ``galaxy.util.resources.resource_path("galaxy.model.migrations", "alembic.ini")`` that should work uniformly across mega-git-Galaxy and standalone packages.

jmchilton commented on lib/galaxy/util/custom_templates/template.py:
```diff
@@ -31,23 +30,9 @@ def _get_template_body(template: str) -> str:
     """Return template file path."""
-    default_path = _get_default_templates_dir() / relpath
+    default_templates_dir = Path(galaxy_directory()) / "lib" / "galaxy" / "config" / "templates"
+    default_path = default_templates_dir / relpath
```
Not your responsibility at all but if you want to "fix" this for Galaxy packages - I'd move this whole module into app and maybe place it in galaxy/config or something like that and then use resources stuff mentioned below.

jmchilton replied:
And to be able to remove jinja2 as a dependency for galaxy-util 😍

jdavcs replied:
@jmchilton, did I move things into their correct locations? (especially the 'test_dbscripts' and 'test_custom_templates')

jdavcs committed:
Move custom templates into the config package

jdavcs committed:
Use importlib.resources to obtain path

jdavcs committed:
Use resource_path for test_migrations

jdavcs committed:
Use resource_path for test_dbscript

jdavcs committed:
Move test_dbscript into app

JUSTIFICATION:
As per @jmchilton:
The package tests should pretend they don't have galaxy_directory()
available because they're meant to be installed into a Python env.
This test is fundamentally not testing anything about the Galaxy data
package - it is testing scripts placed in Galaxy root directory and that
would not be available in an installed Galaxy environment.

jdavcs committed:
Move custom tempates unit test into app; rm jinja2 package dep from config

jdavcs committed:
Re-add jinja2 to config package requirements

jdavcs committed:
Do not call split on cmd: there may be spaces in path

jdavcs committed:
Skip dbscript tests if running from packages; refactor util function

jmchilton commented on lib/galaxy/config/template.py:
```diff
@@ -1,18 +1,17 @@
 
-DEFAULT_TEMPLATES_DIR = Path("lib/galaxy/config/templates")
-PACKAGES_DEFAULT_TEMPLATES_DIR = Path("../config/galaxy/config/templates")
+from galaxy.util.resources import resource_path
```
Lovely!

jmchilton commented on packages/util/setup.cfg:
```diff
@@ -35,7 +35,6 @@ install_requires =
     boltons
     docutils
     importlib_resources
-    jinja2
```
Wonderful!

jmchilton reviewed (approved):
Really nice step forward for packaged Galaxy!

nsoranzo commented on lib/galaxy/config/template.py:
```diff
@@ -1,18 +1,17 @@
 """Render customizable templates e.g. emails with Jinja.
```
Nitpicking, but maybe this filename should be `templates.py` instead of `template.py` ?

jdavcs committed:
Normalize path before checking basename

jdavcs committed:
Rename template.py >> templates.py

neoformit commented on test/unit/data/model/migrations/test_migrations.py:
```diff
@@ -1122,13 +1123,7 @@ def db_state6_gxy_state3_tsi_no_sam(url_factory, metadata_state6_gxy_state3_tsi_
-            path = os.path.join(path, os.pardir)
-        path = os.path.join(path, "lib", "galaxy", "model", "migrations", "alembic.ini")
-
+        path = resource_path("galaxy.model.migrations", "alembic.ini")
```
Much cleaner :grin:

neoformit commented on lib/galaxy/config/templates.py:
```diff
@@ -31,23 +30,8 @@ def _get_template_body(template: str) -> str:
 def _get_template_path(relpath: str, custom_templates_dir: str) -> Path:
     """Return template file path."""
-    default_path = _get_default_templates_dir() / relpath
+    default_path = resource_path("galaxy.config", "templates") / relpath
```
Brilliant

neoformit reviewed (approved):
Looks like a big improvement, I will likely reference this PR in future!
